Senaite labels

ramonski opened the pull request:
## Description of the issue/feature this PR addresses

This PR introduces labels for SENAITE:

<img width="1838" alt="Labels — SENAITE LIMS 2023-03-17 12 PM-38-19" src="https://user-images.githubusercontent.com/713193/225894414-75dc714b-f2ac-4c1b-b7bc-fecaf70382f6.png">

Basically *all* AT/DX contents can be configured to have labels via the SENAITE registry, unless they define their schema solely over a behavior, e.g. a `DataBox`.
In this case, the extended label field does not appear in the edit view, but the contents can be still labeled programmatically with the provided Label API.

<img width="504" alt="OK — SENAITE LIMS 2023-03-17 12 PM-41-44" src="https://user-images.githubusercontent.com/713193/225894992-a663e43d-30a9-466d-8d81-fa14d22f4ae5.png">

The extended label field queries existing labels that have been created in the "Labels" setup folder, but allow also to enter arbitrary labels for a specific object.

Object labels are stored in the annotation storage to avoid attribute clashes and define extend the following field to the object:

- AT Content Types: `ExtLabels`
- DX Content Types: `ext_labels`

When a label is assigned to an object, the object will be cataloged in the `senaite_catalog_label` catalog.

The `labels` index allows to search for all labeled objects in SENAITE.

## Current behavior before PR

No labels available

## Desired behavior after PR is merged

All objects can be labeled

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

ramonski committed:
Implemented Label contents

ramonski committed:
Added upgrade step

ramonski committed:
Show new setup items in view

ramonski committed:
Attached Workflows

ramonski committed:
Default listing view added

ramonski committed:
Setup catalog

ramonski committed:
Show labels categorized

ramonski committed:
Added category getter/setter

ramonski committed:
Started label API

ramonski committed:
Added API test for string checks

ramonski committed:
Removed

ramonski committed:
Changed new function names

ramonski committed:
Extended Label API

ramonski committed:
Removed

ramonski committed:
Schema extender

ramonski committed:
Merge branch '2.x' into senaite-labels

ramonski committed:
Simplify object labels

ramonski committed:
Ensure we have the full object

ramonski committed:
Merge branch '2.x' into senaite-labels

ramonski committed:
Merge branch '2.x' into senaite-labels

ramonski committed:
Add setter for object labels

ramonski committed:
check for object

ramonski committed:
Correctly remove labels

ramonski committed:
AT extender

ramonski committed:
Context property lookup by own key

ramonski committed:
Render label of widget

ramonski committed:
View macro

ramonski committed:
Custom field

ramonski committed:
Whitespace only

ramonski committed:
Add only filled labels

ramonski committed:
Filter empty labels

ramonski committed:
Added catalog setup

ramonski committed:
Implemented search API and index

ramonski committed:
Reindex

ramonski committed:
Added test

ramonski committed:
Implemented label catalog

ramonski committed:
Added listing view for all services

ramonski committed:
Added fulltext index

ramonski committed:
Description added

ramonski committed:
Sort columns

ramonski committed:
Description added

ramonski committed:
Moved field name to config

ramonski committed:
Property lookup for DX Query Select Widget

ramonski committed:
Dynamically add label behavior

ramonski committed:
Added API for enable/disable schema extension

ramonski committed:
display template

ramonski committed:
Added registry settings for labels

ramonski committed:
AT Queryselect improvements

ramonski committed:
Added subscriber for registry field

ramonski committed:
Inherit from ICanHaveLabels

ramonski committed:
Cleanup

ramonski committed:
Changelog updated

ramonski committed:
Merge branch '2.x' into senaite-labels

ramonski committed:
Moved code to label API

ramonski committed:
Initialize labeled types

ramonski committed:
Handle empty labels for AT types

ramonski committed:
Updated description

ramonski committed:
Docstrings added

ramonski committed:
Fixed Comment

xispa commented on src/senaite/core/api/label.py:
```diff
@@ -0,0 +1,277 @@
+    """
+    brains = query_labels(inactive=inactive, **kw)
+    labels = map(api.get_title, brains)
+    return list(set(labels))
```
`kw` may contain a criteria for sorting that will get lost because of `set(labels)`

ramonski replied:
Good catch! Fixed in c4fefdae8

ramonski replied:
Btw: I'm still wrapping the `map` into a `list`, because in Python 3 it is a generator and not a list per default

xispa commented on src/senaite/core/api/label.py:
```diff
@@ -0,0 +1,277 @@
+    :returns: True if the object is a label
+    """
+    if not api.is_object(obj):
+        return False
```
I would skip the first check and simply do:

```python
def is_label_object(obj)
    obj = api.get_object(obj, default=None)
    return ILabel.providedBy(obj)
```

ramonski replied:
Changed in 9b232056b

xispa commented on src/senaite/core/api/label.py:
```diff
@@ -0,0 +1,277 @@
+    if not IHaveLabels.providedBy(obj):
+        return tuple()
+    labels = get_storage(obj)
+    return labels
```
Will return `None` if no labels have been assigned yet to this object. I think  `labels = get_storage(obj, default=tuple())` would be better. `

ramonski replied:
Good catch again! Fixed in a6f47e0fb

ramonski replied:
Well, actually it *should* not happen if `IHaveLabels` is marked. But better to code defensive

xispa commented on src/senaite/core/api/label.py:
```diff
@@ -0,0 +1,277 @@
+    """
+    obj = get_object(obj)
+    if not IHaveLabels.providedBy(obj):
+        return tuple()
```
Maybe this object had the `IHaveLabels` interface in the past and still have labels assigned. With this if, we can end up with a mismatch between what this function returns and what is actually stored and indexed in the catalog if the user removes the interface or add labels manually without using the `set_obj_labels` function. I suggest to skip this check

ramonski replied:
Ok, got your point now. It is true what you say but probably no code in the world will save our day if someone manipulates the data or does not use the proper API functions...
Changed in dd394ca81

xispa commented on src/senaite/core/browser/label/configure.zcml:
```diff
@@ -0,0 +1,13 @@
+
+  <browser:page
+      for="senaite.core.content.labels.ILabels"
+      name="labeled_objects"
```
I think there is no link anywhere for this endpoint. Is this intended? If not, a link next to the [+Add] button at the top of the labels listing would work.

xispa replied:
Note for future: would be nice to make the `senaite.app.listing` to automagically display the column `Labels` if the class of the portal type of the items to be listed implements `ICanHaveLabels`

ramonski replied:
The link is located in the Label Control Panel View:

<img width="1284" src="https://user-images.githubusercontent.com/713193/227030167-44d8b8c7-2ddd-4b6a-a97a-5645fe0c09b3.png">

xispa commented on src/senaite/core/extender/label.py:
```diff
@@ -0,0 +1,94 @@
+    """Extend schema fields for labeled contents
+    """
+    layer = ISenaiteCore
+    implements(
```
As far as I am concerned, the decorators `@implementer` and `@adapter` are preferred nowadays.

ramonski replied:
Yes, that's true. But for multiple implemented Interfaces I prefer this notation

xispa commented on src/senaite/core/api/label.py:
```diff
@@ -0,0 +1,277 @@
+    :param portal_type: The portal_type to enable labeling
+    """
+    klass = get_klass(portal_type)
+    classImplements(klass, ICanHaveLabels)
```
For AT types, this works properly, but for DX types you need to programmatically add the behavior. Something like follows:

```python
behavior_id = 'senaite.core.behavior.labels'
pt = api.get_tool("portal_types")
fti = pt.get(portal_type)
if behavior_id not in fti.behaviors:
    fti_behaviors = list(fti_behaviors)
    fti_behaviors.append(behavior_id)
    fti.behaviors = tuple(fti_behaviors)
```

ramonski replied:
You might be right and I thought too complicated.
I actually implemented this adapter to hook in the extended schema if the object implements `ICanHaveLabels`:
https://github.com/senaite/senaite.core/blob/8259ed672392308b911a4008a2c1cb06b741e34b/src/senaite/core/behaviors/label.py#L25

My initial thought was, that only by marking the class I can apply he behavior to both existing and new objects accordingly.
But the class interface is not looked up when a new DX type is added over the UI and seems to inherit then the interface in the object, which seems to make it not possible to disable it again w/o touching every object.
But adding/removing the behavior to the FTI programmatically might just solve the issue. Let me check this...

xispa reviewed (changes requested):
This is going to be awesome!, I've left some comments

ramonski committed:
Removed `set` when listing global labels

This might contain then duplicates, but at least it won't break the
sorting (if given in **kw).

ramonski committed:
Improved API function

ramonski committed:
Ensure get_obj_labels returns a tuple

ramonski committed:
Always return the labels of the storage

ramonski committed:
Fixed doctest

ramonski committed:
Moved functions into main API

ramonski committed:
Moved behavior functions into main API

ramonski committed:
Notify schema invalidation

ramonski committed:
Handle `None` IDs

ramonski committed:
Treat object w/o ID as temporary

ramonski committed:
Removed custom behavior assignable adapter

ramonski committed:
Removed Annotation Storage

ramonski committed:
No indirect schema extension

ramonski committed:
Enable dexterity behavior

ramonski committed:
Keep behavior assignable adapter

ramonski committed:
Removed confusing tests

ramonski committed:
Renamed method

ramonski committed:
Handle catalog clear&rebuild

ramonski committed:
Convenience getter/setter

ramonski committed:
Added TYPES

ramonski committed:
Avoid to access the object

ramonski committed:
Renamed DX field to match AT label field

ramonski committed:
Metadata added

ramonski committed:
Commented metadata column

ramonski committed:
Renamed fields to "Labels"

ramonski committed:
Cleanup

ramonski committed:
removed metadata column

ramonski committed:
Added getLabels patch for metadata column
